x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/BucketCorrelationAggregationIT.java

```diff
@@ -0,0 +1,228 @@
+        List<Double> percentiles = new ArrayList<>();
+        List<Double> fractions = new ArrayList<>();
+        RangeAggregationBuilder builder = AggregationBuilders.range("correlation_range").field(indicatorFieldName);
+        double percentile_0 = raw_percentiles.percentile(steps[0]);
+        builder.addUnboundedTo(percentile_0);
+        fractions.add(0.02);
+        percentiles.add(percentile_0);
+        int last_added = 0;
+        for (int i = 1; i < steps.length; i++) {
+            double percentile_l = raw_percentiles.percentile(steps[i - 1]);
+            double percentile_r = raw_percentiles.percentile(steps[i]);
+            if (Double.compare(percentile_l, percentile_r) == 0) {
+                fractions.set(last_added, fractions.get(last_added) + 0.02);
+            } else {
+                last_added = i;
+                fractions.add(0.02);
+                percentiles.add(percentile_r);
+            }
+        }
+        fractions.add(2.0/100);
+        double[] expectations = new double[percentiles.size() + 1];
+        expectations[0] = percentile_0;
+        for (int i = 1; i < percentiles.size(); i++) {
+            double percentile_l = percentiles.get(i - 1);
+            double percentile_r = percentiles.get(i);
+            double fractions_l = fractions.get(i - 1);
+            double fractions_r = fractions.get(i);
+            builder.addRange(percentile_l, percentile_r);
+            expectations[i] = (fractions_l * percentile_l + fractions_r * percentile_r) / (fractions_l + fractions_r);
+        }
+        double percentile_n = percentiles.get(percentiles.size() - 1);
+        builder.addUnboundedFrom(percentile_n);
+        expectations[percentiles.size()] = percentile_n;
+        return Tuple.tuple(
+            builder,
+            new BucketCorrelationAggregationBuilder(
+                "correlates",
+                "correlation_range>_count",
+                new CountCorrelationFunction(
+                    new CorrelativeValue(expectations, fractions.stream().mapToDouble(Double::doubleValue).toArray(), totalCount)
+                )
+            )
+        );
+    }
+
+    private double pearsonCorrelation(double[] xs, int[] ys) {
+        double meanX = sum(xs)/xs.length;
+        double meanY = sum(ys)/(double)ys.length;
+        double varX = 0.0;
+        for (double x : xs) {
+            varX += Math.pow(x - meanX, 2);
+        }
+        varX /= xs.length;
+
+        double varY = 0.0;
+        for (int y : ys) {
+            varY += Math.pow(y - meanY, 2);
+        }
+        varY /= ys.length;
+
+        if (varY == 0 || varX == 0) {
+            fail("failed to calculate true correlation due to 0 variance in the data");
+        }
+
+        double corXY = 0.0;
+        for (int i = 0; i < xs.length; i++) {
+            corXY += (((xs[i] - meanX)*(ys[i] - meanY))/Math.sqrt(varX*varY));
+        }
+        return corXY/xs.length;
+    }
+
+    private static double sum(double[] xs) {
```

Comment:
Is `DoubleStream.of(xs).sum()` short enough that you wouldn't need to make a whole function for it? Its up to you.

Reply:
I am simply using `MovingFunctions.sum` now in a subsequent commit.